python/ml4ir/base/features/feature_fns/categorical.py

```diff
@@ -0,0 +1,260 @@
+
+    categorical_fc = feature_column.categorical_column_with_identity(
+        CATEGORICAL_VARIABLE,
+        num_buckets=feature_layer_info["args"]["num_buckets"],
+        default_value=feature_layer_info["args"].get("default_value", None),
+    )
+    embedding_fc = feature_column.embedding_column(
+        categorical_fc, dimension=feature_layer_info["args"]["embedding_size"]
+    )
+
+    embedding = layers.DenseFeatures(
+        embedding_fc,
+        name="{}_embedding".format(feature_info.get("node_name", feature_info["name"])),
+    )({CATEGORICAL_VARIABLE: feature_tensor})
+    embedding = tf.expand_dims(embedding, axis=1)
+
+    return embedding
+
+
+class VocabLookup(layers.Layer):
+    """
+    Class defines a keras layer wrapper around a tf lookup table using the given vocabulary list.
+    Maps each entry of a vocabulary list into categorical indices.
+
+    Attributes:
+        vocabulary_list: List of strings that form the vocabulary set of categorical values
+        num_oov_buckets: Number of buckets to be used for out of vocabulary strings
+        feature_name: Name of the input feature tensor
+        lookup_table: Tensorflow look up table that maps strings to integer indices
+
+    NOTE:
+    Issue[1] with using LookupTable with keras symbolic tensors; expects eager tensors.
+
+    Ref: https://github.com/tensorflow/tensorflow/issues/38305
+    """
+
+    def __init__(self, vocabulary_list, num_oov_buckets, feature_name):
+        super(VocabLookup, self).__init__(trainable=False, dtype=tf.int64)
+        self.vocabulary_list = vocabulary_list
+        self.vocabulary_size = len(vocabulary_list)
+        self.num_oov_buckets = num_oov_buckets
+        self.feature_name = feature_name
+
+    def build(self, input_shape):
+        table_init = lookup.KeyValueTensorInitializer(
+            keys=self.vocabulary_list,
+            values=range(self.vocabulary_size),
+            key_dtype=tf.string,
+            value_dtype=tf.int64,
+        )
+        self.lookup_table = lookup.StaticVocabularyTable(
+            initializer=table_init,
+            num_oov_buckets=self.num_oov_buckets,
+            name="{}_lookup_table".format(self.feature_name),
+        )
+        self.built = True
+
+    def call(self, input_text):
+        return self.lookup_table.lookup(input_text)
+
+    def get_config(self):
+        config = super(VocabLookup, self).get_config()
+        config.update(
+            {
+                "vocabulary_list": self.vocabulary_list,
+                "vocabulary_size": self.vocabulary_size,
+                "num_oov_buckets": self.num_oov_buckets,
+                "feature_name": self.feature_name,
+            }
+        )
+        return config
+
+
+def categorical_embedding_with_vocabulary_file(feature_tensor, feature_info):
+    """
+    Converts a string tensor into a categorical embedding representation.
+    Works by using a vocabulary file to convert the string tensor into categorical indices
```

Comment:
Also in the class definition above: Do we convert a string to explicitly one categorical index or several categorical indices or both are possible?

Reply:
It is one categorical index. So a 1->1. Alternatively, what I can do is read a file with the key and the index(value) mapping. That way we can make it many to one. Should we do that?

Reply:
I tested the many to one vocabulary lookup. It works.